examples/inspection/plot_permutation_importance_SHAP.py

```diff
@@ -0,0 +1,336 @@
+# the Shapley values. The linear explanation model equation is:
+#
+# .. math::
+#   g(z')=\phi_0 + \sum_{i=1}^{M} \phi_i z'
+#
+# where :math:`g(z')` is the explanatory model, :math:`\phi_0` is the model
+# prediction when all features are 'missing', :math:`M` is the number of
+# possible subset sizes (n_features - 1), :math:`z'\in\{0,1\}^M`
+# (denotes the presence or absence of each feature) and :math:`\phi_i` is the
+# estimated Shapley value.
+#
+# This is much more computationally expensive than permutation importance
+# particularly as the number of possible combinations of features for all
+# feature subsets sizes quickly becomes very large as the number of features
+# increases. However, this does enable SHAP to account for interaction effects
+# between features.
+#
+# First, we will instantiate ``KernelSHAP`` using our fitted regressor,
+# ``reg`` and some background data, which will be used to simulate missingness.
+# If your dataset is small (e.g., training data is <100 samples) you can use
+# the whole training subset for the background data. Our data is larger than
+# this so we must summarize it in some way, otherwise computation will be
+# too slow. For simplicity, we will use the
+# median values of our features but SHAP offers a ``kmeans`` function that
+# can summarize each feature with ``k`` means.
+
+import shap
+
+med = X_train.median().values.reshape((1, X_train.shape[1]))
+explainer = shap.KernelExplainer(reg.predict, med)
+
+# %%
+# ``explainer`` stores various information about the data as attributes. Of
+# interest is ``expected_value``. This represents :math:`\phi_0` in our
+# equation above.
+
+explainer.expected_value
+
+# %%
+# Can you work out how this value was calculated?
+#
+# It is the prediction output by our model when using the background
+# values we gave it.
+
+reg.predict(med)
+
+# %%
+# Next we will calculate Shapley values using the testing subset. This is the
+# computationally expensive step.
+
+shap_values = explainer.shap_values(X_test, l1_reg="aic")
+
+# %%
+# Let's look at the Shapley values of one sample. There are 8 Shapley values,
+# one for each feature. '0' means that the feature did not contribute to the
+# prediction output. A negative value 'pushes' the prediction lower and a
+# positive value pushes the prediction higher.
+
+shap_values[0, :]
+
+# %%
+# The Shapley values should sum to the difference between the
+# prediction output by our our model ``reg`` and  the ``expected_value`
+# (depending on how well the linear model was able to be fit).
+
+print(f"The sum of Shapley values: {shap_values[0, :].sum()}")
+prediction = reg.predict(X_test.iloc[0, :].values.reshape(1, -1))[0]
+print(
+    "Difference between prediction and expected value: "
+    f"{prediction - explainer.expected_value}"
+)
+
+# %%
+# We can also plot the Shapley values:
+
+import matplotlib.pyplot as plt
+
+shap.summary_plot(shap_values, X_test, show=False)
```

Comment:
Since there is no colorbar, it should be nice to explain what blue/red are representing.

Reply:
Locally the color bar appears but have updated to remove some redundant things so maybe it will fix

Reply:
The color bar is wrong locally too now and I've tried a few versions of shap (going back to v0.30) and cannot make the color bar correct. I think it is this that causes the labels to show in the middle and without the bar:
https://github.com/slundberg/shap/blob/9411b68e8057a6c6f3621765b89b24d82bee13d4/shap/plots/_beeswarm.py#L870-L871

Since the function does not return anything, I am not sure I can fix it.

I've tried using their new API but the plot looks the same and it does not seem to support kernel explainer atm (even though the docstring says it does).

I could create an issue in shap and see what happens (though question the stability of the plotting) or I can just us the values to create my own plot..?

Reply:
Indeed I was also confused by the fact that the generic `Explainer` class does not provide the kernel method, despite the docstring saying so for the `algorithm` parameter. Feel free to open an issue upstream to ask about this usability problem.